### ISSUE TITLE
Support the ?? operator for Code Path Analysis

### ISSUE THREAD

**Issue by gaearon** [eslint/eslint#12140]
The [React Hooks lint rule](https://www.npmjs.com/package/eslint-plugin-react-hooks) uses [Code Path Analysis](https://eslint.org/docs/2.0.0/developer-guide/code-path-analysis) to analyze code path segments. Specifically, we verify that React Hooks aren't called conditionally.

The rule successfully flags this code as a violation:

```js
function MyComponent({ bar }) {
  const foo = bar && useFoo();
}
```

But not this:

```js
function MyComponent({ bar }) {
  const foo = bar ?? useFoo();
}
```

I suppose this is because `??` support is rudimentary (#10475) and so we don't actually include it in the Code Path Analysis. Now that `??` is Stage 3, can we revisit this? Thanks.

**Comment by kaicataldo** [comment]
Our [policy](https://github.com/eslint/eslint#what-about-experimental-features) states that we only officially support language features that reach stage 4 (and that we'll update core to not crash on stage 3 syntax), since we allow for custom parsers to handle that syntax (like babel-eslint). We generally don't make exceptions to this (we used to, which led to us codifying our policy). 

That being said, I don't think there's a way to do this currently, and I think it's worth discussing and seeing how one might accomplish this, whether that's creating an API for the extension of code path analysis or some other solution.
